File: database/conexao.php
<?php
/**
 * CashPilot - Conexão PDO.
 *
 * Local/XAMPP:
 *   DB_HOST=localhost, DB_NAME=cashpilot, DB_USER=root, DB_PASS=""
 *
 * Railway:
 *   aceita diretamente MYSQLHOST, MYSQLPORT, MYSQLDATABASE,
 *   MYSQLUSER e MYSQLPASSWORD fornecidas pelo serviço MySQL.
 */

function cashpilotEnvPrimeiro(array $nomes, string $padrao = ''): string
{
    foreach ($nomes as $nome) {
        $valor = getenv($nome);
        if (is_string($valor) && trim($valor) !== '') {
            return trim($valor);
        }
    }

    return $padrao;
}

function conectar(): PDO
{
    static $pdo = null;

    if ($pdo instanceof PDO) {
        return $pdo;
    }

    $host = cashpilotEnvPrimeiro(['DB_HOST', 'MYSQLHOST'], 'localhost');
    $porta = cashpilotEnvPrimeiro(['DB_PORT', 'MYSQLPORT'], '3306');
    $nome = cashpilotEnvPrimeiro(['DB_NAME', 'MYSQLDATABASE'], 'cashpilot');
    $usuario = cashpilotEnvPrimeiro(['DB_USER', 'MYSQLUSER'], 'root');
    $senha = cashpilotEnvPrimeiro(['DB_PASS', 'MYSQLPASSWORD'], '');

    try {
        $dsn = 'mysql:host=' . $host
            . ';port=' . (int) $porta
            . ';dbname=' . $nome
            . ';charset=utf8mb4';

        $pdo = new PDO($dsn, $usuario, $senha, [
            PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
            PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
            PDO::ATTR_EMULATE_PREPARES => false,
            PDO::ATTR_STRINGIFY_FETCHES => false,
        ]);

        return $pdo;
    } catch (PDOException $e) {
        error_log('CashPilot/DB: ' . $e->getMessage());
        http_response_code(500);
        exit('Não foi possível conectar ao CashPilot neste momento.');
    }
}

File: python/conexao.py
"""
CashPilot - conexão Python com MySQL.

Aceita tanto as variáveis DB_* usadas pelo projeto quanto as variáveis
MYSQL* fornecidas automaticamente por provedores como Railway.
"""

import os

import mysql.connector


def _env(*names: str, default: str = "") -> str:
    for name in names:
        value = os.getenv(name)
        if value is not None and value.strip() != "":
            return value.strip()
    return default


def conectar():
    """Abre e retorna uma conexão com o banco de dados MySQL."""
    config = {
        "host": _env("DB_HOST", "MYSQLHOST", default="localhost"),
        "port": int(_env("DB_PORT", "MYSQLPORT", default="3306")),
        "user": _env("DB_USER", "MYSQLUSER", default="root"),
        "password": _env("DB_PASS", "MYSQLPASSWORD", default=""),
        "database": _env("DB_NAME", "MYSQLDATABASE", default="cashpilot"),
        "charset": "utf8mb4",
        "use_unicode": True,
    }

    return mysql.connector.connect(**config)
